### PR TITLE
Changed LSP completion resolve to not cache data on items.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- We used to cache displayText, text document, position and completion triggers on the data field of a completion item however in practice this resulted in having a pretty massive footprint on how completion lists serialized/deserialized.
  - Because completion resolve used to cache more data and now we don't I added a telemetry event to capture how often a cache miss actually happens
  - As part of this I had to migrate the `CompletionListCache` to be synchronous so that we can synchronously resolve cache entries at completion resolve time.
- Updated tests to reflect new resolve behavior. As part of doing this I had to expand the completion resolve tests to actually generate completion lists so caches were properly updated. Ultimately this more accurately reflects what happens in real scenarios.
- Wanted to use records so added an `IsExternalInit.cs` file to enable record usage in netstandard2.0.

## Benchmark results

**Scenario:** Typing `@` in a Razor file and getting a basic C# completion list.

Optimized = true represents utilizing the Optimized completion list (which Roslyn does now) and unoptimized represents anyone who's consuming Roslyn's completion list who doesn't use the optimized completion list structure.

### Before (817KB)
|                                                    Method | Optimized |      Mean |     Error |    StdDev |   Op/s |     Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|---------------------------------------------------------- |---------- |----------:|----------:|----------:|-------:|----------:|---------:|---------:|----------:|
| 'Completion List Roundtrip Serialization/Deserialization' |     False | 29.301 ms | 0.5816 ms | 0.6698 ms |  34.13 | 1718.7500 | 937.5000 | 250.0000 |    9.8 MB |
|                           'Completion List Serialization' |     False | 10.223 ms | 0.1926 ms | 0.1608 ms |  97.82 |  625.0000 | 250.0000 | 234.3750 |   3.27 MB |
|                         'Completion List Deserialization' |     False | 18.653 ms | 0.3708 ms | 0.5882 ms |  53.61 | 1062.5000 | 531.2500 |        - |   6.52 MB |
| 'Completion List Roundtrip Serialization/Deserialization' |      True | 19.509 ms | 0.3308 ms | 0.3094 ms |  51.26 | 1218.7500 | 718.7500 | 218.7500 |   7.03 MB |
|                           'Completion List Serialization' |      True |  2.121 ms | 0.0259 ms | 0.0216 ms | 471.46 |  199.2188 | 167.9688 | 164.0625 |    1.1 MB |
|                         'Completion List Deserialization' |      True | 18.869 ms | 0.3342 ms | 0.3849 ms |  53.00 | 1062.5000 | 531.2500 |        - |   6.52 MB |

### After (458KB)
|                                                    Method | Optimized |      Mean |     Error |    StdDev |   Op/s |     Gen 0 |    Gen 1 |    Gen 2 |  Allocated |
|---------------------------------------------------------- |---------- |----------:|----------:|----------:|-------:|----------:|---------:|---------:|-----------:|
| 'Completion List Roundtrip Serialization/Deserialization' |     False | 19.438 ms | 0.3835 ms | 0.3400 ms |  51.45 | 1093.7500 | 406.2500 | 125.0000 | 7216.06 KB |
|                           'Completion List Serialization' |     False |  8.355 ms | 0.0735 ms | 0.0651 ms | 119.68 |  437.5000 | 171.8750 | 125.0000 | 2839.71 KB |
|                         'Completion List Deserialization' |     False | 10.303 ms | 0.2045 ms | 0.5701 ms |  97.06 |  625.0000 | 312.5000 |        - | 3844.78 KB |
| 'Completion List Roundtrip Serialization/Deserialization' |      True | 10.847 ms | 0.2057 ms | 0.2021 ms |  92.20 |  656.2500 | 234.3750 | 109.3750 | 4458.84 KB |
|                           'Completion List Serialization' |      True |  1.163 ms | 0.0064 ms | 0.0060 ms | 860.16 |  123.0469 | 123.0469 | 123.0469 |  614.06 KB |
|                         'Completion List Deserialization' |      True | 10.103 ms | 0.2004 ms | 0.3909 ms |  98.98 |  625.0000 | 312.5000 |        - | 3844.78 KB |

### Results (Optimized = true)

**Roundtrip Improvement:** 1.8x
**Serialization Improvement:** 1.82x
**Deserialization Improvement:** 1.87x

Part of dotnet/aspnetcore#30232

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs [hunk 1]
```diff
@@ -4,9 +4,7 @@
 
 #nullable disable
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:5]: if it's not too impactful, could you remove this?

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs [hunk 2]
```diff
@@ -50,35 +64,17 @@ public async Task<LSP.CompletionItem> HandleRequestAsync(LSP.CompletionItem comp
...
-            CompletionList? list = null;
-
-            // See if we have a cache of the completion list we need
-            if (data.ResultId.HasValue)
+            var cacheEntry = GetCompletionListCacheEntry(completionItem);
+            if (cacheEntry == null)
             {
-                list = await _completionListCache.GetCachedCompletionListAsync(data.ResultId.Value, cancellationToken).ConfigureAwait(false);
+                // Don't have a cache associated with this completion item, cannot resolve.
+                return completionItem;
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:71]: Even though we log telemetry already, we should also log a loghub warning here.  If we get a feedback about a completion error it's really easy to tell exactly which completion request hit this (which we couldn't do by just looking at the telemetry).

File: src/Features/LanguageServer/Protocol/IsExternalInit.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Runtime.CompilerServices
+{
+    // Used to compile against C# 9 in a netstandard2.0
+    internal class IsExternalInit
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/IsExternalInit.cs:8]: @sharwell do we already have this somewhere / should this be moved somewhere else?

File: src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionResolveTests.cs [hunk 4]
```diff
@@ -106,11 +116,15 @@ class B : A
...
+            var serverCompletionItem = completionList.Items.FirstOrDefault(item => item.Label == "M()");
+            var completionResultId = ((CompletionResolveData)serverCompletionItem.Data).ResultId.Value;
             var document = testLspServer.GetCurrentSolution().Projects.First().Documents.First();
 
-            var completionItem = await CreateCompletionItemAsync(
-                label: "M()", LSP.CompletionItemKind.Method, tags, completionParams, document,
-                commitCharacters: CompletionRules.Default.DefaultCommitCharacters).ConfigureAwait(false);
+            // Reconstruct the completion item in our view like a client would while maintaining data
+            var clientCompletionItem = await CreateCompletionItemAsync(
+                label: "M()", kind: LSP.CompletionItemKind.Method, tags: tags, request: completionParams, document: document,
```
Comment (dibarbet) [inline src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionResolveTests.cs:126]: could we extract this out to a common helper?  Looks like this mostly similar code is repeated above

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs [hunk 5]
```diff
@@ -31,15 +33,27 @@ public CompletionResolveHandler(CompletionListCache completionListCache)
...
+            {
+                return null;
+            }
 
-            return ((JToken)request.Data).ToObject<CompletionResolveData>();
+            var cacheEntry = _completionListCache.GetCachedCompletionList(resolveData.ResultId.Value);
+            if (cacheEntry == null)
+            {
+                // No cache for associated completion item. Log some telemetry so we can understand how frequently this actually happens.
+                Logger.Log(FunctionId.LSP_CompletionListCacheMiss, KeyValueLogMessage.NoProperty);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:49]: Even though we log telemetry already, we should also log a loghub warning here.  If we get a feedback about a completion error it's really easy to tell exactly which completion request hit this (which we couldn't do by just looking at the telemetry).

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs [hunk 6]
```diff
@@ -31,15 +33,27 @@ public CompletionResolveHandler(CompletionListCache completionListCache)
             _completionListCache = completionListCache;
         }
 
-        private static CompletionResolveData GetCompletionResolveData(LSP.CompletionItem request)
+        private CompletionListCache.CacheEntry? GetCompletionListCacheEntry(LSP.CompletionItem request)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:36]: nit: prefer ordering private methods below the public / overriden methods

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs [hunk 7]
```diff
@@ -31,15 +33,27 @@ public CompletionResolveHandler(CompletionListCache completionListCache)
...
         {
             Contract.ThrowIfNull(request.Data);
+            var resolveData = ((JToken)request.Data).ToObject<CompletionResolveData>();
+            if (resolveData?.ResultId == null)
+            {
+                return null;
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:42]: If ResultId is null, should we throw here instead via Contract check?

**Review by allisonchou — APPROVED** [review]
lgtm!

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 1 above)
Comment (NTaylorMullen) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:5]: It complains about the underlying `TextDocumentIdentifier` below

Comment (NTaylorMullen) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:5]: I'm an idiot, I removed the TextDocumentIdentifier which was outside of our control and misremembered. YOu are 100% right

File: src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionResolveTests.cs (hunk 4 above)
Comment (NTaylorMullen) [reply src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionResolveTests.cs]: Created a `ConvertToClientCompletionItem` method that serializes then deserializes the server item to more accurately replicate what happens in practice

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs [hunk 8]
```diff
@@ -2,11 +2,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+#nullable enable
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:6]:
```suggestion
```

all projects should be nullable enabled by default

**Review by dibarbet — APPROVED** [review]

**Pull request merged**